aiida/manage/configuration/migrations/migrations.py

```diff
@@ -8,136 +8,299 @@
+
+    down_revision: int
+    """The initial configuration version."""
 
-    def __init__(self, migrate_function, version, version_oldest_compatible):
-        """Construct a ConfigMigration
+    down_compatible: int
+    """The initial oldest backwards compatible configuration version"""
 
-        :param migrate_function: function which migrates the configuration dictionary
-        :param version: configuration version after the migration.
-        :param version_oldest_compatible: oldest compatible configuration version after the migration.
-        """
-        self.migrate_function = migrate_function
-        self.version = int(version)
-        self.version_oldest_compatible = int(version_oldest_compatible)
+    up_revision: int
+    """The final configuration version."""
 
-    def apply(self, config):
-        """Apply the migration to the configuration."""
-        config = self.migrate_function(config)
-        config.setdefault('CONFIG_VERSION', {})['CURRENT'] = self.version
-        config.setdefault('CONFIG_VERSION', {})['OLDEST_COMPATIBLE'] = self.version_oldest_compatible
-        return config
+    up_compatible: int
+    """The final oldest backwards compatible configuration version"""
 
+    def upgrade(self, config: ConfigType) -> None:
+        """Migrate the configuration in-place."""
 
-def _1_add_profile_uuid(config):
+    def downgrade(self, config: ConfigType) -> None:
+        """Downgrade the configuration in-place."""
+
+
+class Initial(SingleMigration):
     """Add the required values for a new default profile.
 
         * PROFILE_UUID
 
     The profile uuid will be used as a general purpose identifier for the profile, in
     for example the RabbitMQ message queues and exchanges.
     """
-    for profile in config.get('profiles', {}).values():
-        from uuid import uuid4
-        profile['PROFILE_UUID'] = uuid4().hex
+    down_revision = 0
+    down_compatible = 0
+    up_revision = 1
+    up_compatible = 0
 
-    return config
+    def upgrade(self, config: ConfigType) -> None:
+        pass
+
+    def downgrade(self, config: ConfigType) -> None:
+        pass
 
 
-def _2_simplify_default_profiles(config):
+class AddProfileUuid(SingleMigration):
+    """Add the required values for a new default profile.
+
+        * PROFILE_UUID
+
+    The profile uuid will be used as a general purpose identifier for the profile, in
+    for example the RabbitMQ message queues and exchanges.
+    """
+    down_revision = 1
+    down_compatible = 0
+    up_revision = 2
+    up_compatible = 0
+
+    def upgrade(self, config: ConfigType) -> None:
+        from uuid import uuid4  # we require this import here, to patch it in the tests
+        for profile in config.get('profiles', {}).values():
+            profile.setdefault('PROFILE_UUID', uuid4().hex)
+
+    def downgrade(self, config: ConfigType) -> None:
+        # leave the uuid present, so we could migrate back up
+        pass
```

Comment:
```suggestion
```

Reply:
you can't leave just a comment (and the docstring is being inherited)